### PR TITLE
Task/acc 2905/service branch

### PR DESCRIPTION
**Pull request by l0rdcafe**

### CODE REVIEW

File: cli/commands/up.go [hunk 1]
```diff
@@ -12,12 +13,13 @@ import (
...
-	skipLazydocker    bool
-	playlistName      string
-	serviceNames      []string
+	skipServicePreRun  bool
+	skipGitPull        bool
+	skipDockerPull     bool
+	skipLazydocker     bool
+	playlistName       string
+	serviceNames       []string
+	serviceBranchNames []string
```
Comment (l0rdcafe) [inline cli/commands/up.go:22]: fmt changes + adding `serviceBranchNames`

File: cli/commands/up.go [hunk 2]
```diff
@@ -113,6 +124,7 @@ Run the postgres and localstack services directly:
 	flags.BoolVar(&opts.skipDockerPull, "no-remote-pull", false, "Don't get new remote images")
 	flags.BoolVar(&opts.skipLazydocker, "no-lazydocker", false, "Don't start lazydocker")
 	flags.StringVarP(&opts.playlistName, "playlist", "p", "", "The name of a playlist")
+	flags.StringSliceVarP(&opts.serviceBranchNames, "branch", "b", []string{}, "Comma separated list of service:branch to run")
```
Comment (ohsabry) [inline cli/commands/up.go:127]: Can we make this not need to be comma separated?

File: README.md [hunk 3]
```diff
@@ -180,6 +181,22 @@ Override schema:
...
+sso tb up venue-config-service -t venue-config-service:my_branch
+```
+
+You can also override a remote tag when running a playlist. To override multiple remote tags for multiple services, you will need to provide a comma-separated list of service:tag arguments.
+
+```
+sso tb up -p vcs-deps -t postgres:my_branch1,redis:my_branch2
+```
+
+NOTE: ECR does not support forward slashes (/) in an image tag, so task/ACC-1234/chore will be normalized to task-ACC-1234-chore when the image is pushed to ECR.
```
Comment (ohsabry) [inline README.md:198]: This is touchbistro specific stuff, tb is a more general tool. I would document that slash gotcha in tb-registry

File: README.md [hunk 4]
```diff
@@ -180,6 +181,22 @@ Override schema:
     tag: string      # The image tag to use
 ```
 
+#### Overriding Remote Tag using CLI
+
+Additionally, you can run a docker image with a specific remote tag using the CLI. An example of doing so looks like this:
+
+```
+sso tb up venue-config-service -t venue-config-service:my_branch
```
Comment (ohsabry) [inline README.md:189]: `my_branch` => `my_tag`

Comment (ohsabry) [reply README.md:189]: Can we find a way to not need to duplicate the service name, if you are only running a single service? ie

`sso tb up venue-config-service -t my_tag`

`sso tb up -p vcs-deps -t postgres:m_tag1,redis:my_tag2`

File: cli/commands/up.go [hunk 5]
```diff
@@ -69,13 +71,22 @@ Run the postgres and localstack services directly:
 			if len(serviceNames) == 0 {
 				serviceNames = opts.serviceNames
 			}
+
+			serviceTags := make(map[string]string)
+			for _, serviceTag := range opts.serviceTags {
+				sb := strings.Split(serviceTag, ":")
+				// replacing forward slashes with dashes (because that's how ECR images are tagged)
+				trimmedBranch := strings.ReplaceAll(sb[1], "/", "-")
```
Comment (ohsabry) [inline cli/commands/up.go:79]: This is too internal to how we happen to tag things in ECR in our own CI. I would take this out, and have people use the correct tags directly

File: engine/service.go [hunk 6]
```diff
@@ -562,25 +564,57 @@ func (e *Engine) nuke(ctx context.Context, opts NukeOptions, op errors.Op) error
...
 			services[i] = s
+			tag := serviceTags[s.Name]
+			if len(tag) > 0 {
+				override := service.ServiceOverride{
+					Mode: "remote",
+					Remote: service.RemoteOverride{
+						Tag: tag,
+					},
+				}
+				override.Remote.Tag = tag
```
Comment (ohsabry) [inline engine/service.go:602]: This seems redundant if you're adding the tag while creating the override

File: engine/service.go [hunk 7]
```diff
@@ -562,25 +564,57 @@ func (e *Engine) nuke(ctx context.Context, opts NukeOptions, op errors.Op) error
...
 // is not supported.
 //
 // If neither serviceNames nor playlistName are provided, then behaviour depends on the value of requireOne.
 // In this case, if requireOne is true, an error will be returned since at least one of serviceNames or playlistName
 // was required. Otherwise, both the returned slice and error will be nil, which can be treated as an empty slice
 // of services.
-func (e *Engine) resolveServices(op errors.Op, serviceNames []string, playlistName string, requireOne bool) ([]service.Service, error) {
-	if len(serviceNames) > 0 && playlistName != "" {
+func (e *Engine) resolveServices(op errors.Op, serviceNames []string, playlistName string, serviceTags map[string]string, requireOne bool) ([]service.Service, error) {
+	if len(serviceNames) > 0 && playlistName != "" && len(serviceTags) == 0 {
```
Comment (ohsabry) [inline engine/service.go:575]: Why would we restrict this validation to only the case where we aren't using service tags? even if we have multiple tags, we shouldn't be accepting both a playlist and a list of services

File: engine/service.go (hunk 6 above)
Comment (l0rdcafe) [reply engine/service.go:602]: good catch, yeah we don't need this assignment

File: cli/commands/up.go (hunk 5 above)
Comment (l0rdcafe) [reply cli/commands/up.go:79]: ok cool, i'll remove it then

File: README.md (hunk 4 above)
Comment (l0rdcafe) [reply README.md]: got it working btw

File: engine/service.go (hunk 7 above)
Comment (l0rdcafe) [reply engine/service.go]: you're right, that must have been a relic from when i was trying things out. i removed that extra check

File: cli/commands/up.go [hunk 8]
```diff
@@ -113,6 +128,7 @@ Run the postgres and localstack services directly:
 	flags.BoolVar(&opts.skipDockerPull, "no-remote-pull", false, "Don't get new remote images")
 	flags.BoolVar(&opts.skipLazydocker, "no-lazydocker", false, "Don't start lazydocker")
 	flags.StringVarP(&opts.playlistName, "playlist", "p", "", "The name of a playlist")
+	flags.StringSliceVarP(&opts.serviceTags, "image-tag", "t", []string{}, "Comma separated list of service:image-tag to run")
```
Comment (ohsabry) [inline cli/commands/up.go:131]: should document how its used with services and with playlists?

Comment (l0rdcafe) [reply cli/commands/up.go:150]: i added some docs, lmk if there is anything else you had in mind that we should add

**Review by ohsabry — APPROVED** [review]

**Pull request merged**